### PR TITLE
Remove the C Kconfig tools

### PR DESCRIPTION
**Pull request by ulfalizer**
This patchset removes the C Kconfig tools, switching over to the new `menuconfig.py` menuconfig implementation. It also updates the documentation.

Please scream if you spot something you use being removed.

You can already try out the new `menuconfig` interface by running e.g. `ninja pymenuconfig`.

See the commit messages for more information.

### CODE REVIEW

**Comment by ulfalizer** [comment]
CC @mbolivar @carlescufi

**Review by carlescufi — CHANGES_REQUESTED** [review]
Blocking this until we agree on how to present `windows-curses` as a requirement

File: doc/getting_started/installation_win.rst
Comment (ulfalizer) [reply doc/getting_started/installation_win.rst]: Looks like it's possible: https://stackoverflow.com/a/11704396/4577158

I'll change it.

File: doc/application/application.rst
Comment (carlescufi) [inline doc/application/application.rst]: replace "generate a build system" with "generate all required build files"

File: doc/application/application.rst
Comment (carlescufi) [inline doc/application/application.rst]: replace "commands" with "key bindings"

File: doc/application/application.rst
Comment (carlescufi) [inline doc/application/application.rst]: Why did you remove this paragraph? Seems to still apply and there's valuable info there.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: cmake/kconfig.cmake
Comment (carlescufi) [inline cmake/kconfig.cmake]: I wonder if we need a target list anymore? Could we simplify this? Maybe a follow-up PR?

File: scripts/kconfig/checkconfig.py
Comment (carlescufi) [inline scripts/kconfig/checkconfig.py]: so why do we keep this script here at all?

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: doc/application/application.rst
Comment (ulfalizer) [reply doc/application/application.rst]: What about "generate build files"?

File: doc/getting_started/installation_win.rst
Comment (carlescufi) [reply doc/getting_started/installation_win.rst]: Great, this is the better solution IMHO.

File: scripts/kconfig/checkconfig.py
Comment (ulfalizer) [reply scripts/kconfig/checkconfig.py]: Thought it might be more likely that someone might update it if it's kept. I have no strong opinions here though.

File: doc/application/application.rst
Comment (carlescufi) [reply doc/application/application.rst]: sounds good

File: doc/application/application.rst
Comment (ulfalizer) [reply doc/application/application.rst]: It wasn't removed, just dedented. See below. :)

File: doc/application/application.rst
Comment (ulfalizer) [reply doc/application/application.rst]: Will do

File: doc/application/application.rst
Comment (ulfalizer) [reply doc/application/application.rst]: Ah, sorry, I see which paragraph you mean now.

I wonder how helpful it is in practice to explain how to list files starting with `.` here. The description is *nix-specific as well, as those files are always shown on Windows.

Maybe a short note along the lines of `Note that 'ls' does not list files starting with '.' by default. You can pass the '-a' flag to see them.` would be enough, if we keep it.

**Comment by carlescufi** [comment]
>diffconfig is actually a nice script that we should keep IMO

I agree. Does it need changes to work with the future `rsource` and `gsource` directives?

**Comment by nashif** [comment]
> I agree. Does it need changes to work with the future rsource and gsource directives?

no, this does not use the backend, it just does python based diffing on two configs

**Comment by carlescufi** [comment]
@ulfalizer During the TSC call today I mentioned this PR and there was one objection to get rid of the C tools: search support. Would it be possible to add this to `pymenuconfig` before we merge this PR in order to make sure we do not disturb the workflow of the different developers using Zephyr?

File: doc/getting_started/installation_win.rst
Comment (mbolivar) [reply doc/getting_started/installation_win.rst]: Yeah, this is cool

File: doc/application/application.rst
Comment (mbolivar) [inline doc/application/application.rst]: Perhaps:

```
by editing :file:`zephyr/.config` in the application build directory
```

**Review by mbolivar — COMMENTED** [review]
@ulfalizer BTW, it was my concern at the TSC meeting that before we rip out the existing C implementations, we should at least have a search feature, if not both search and jump to definition.

Without that, this is a feature regression for users of the C tools, and it's no longer a no-op to merge.

**Comment by ulfalizer** [comment]
Morning

Alright, I'll add back `diffconfig`. I removed everything in the PR just to see what people are attached to. :)

Adding search/jump-to before integration is fine by me. I'll start working on it.

File: doc/application/application.rst
Comment (carlescufi) [reply doc/application/application.rst]: yes, maybe something like "The Zephyr configuration file generated during CMake configuration is stored in `zephyr/.config`, and is not visible by default on Linux and macOS. Use `ls -a` to be able to see it"

File: scripts/kconfig/checkconfig.py
Comment (carlescufi) [reply scripts/kconfig/checkconfig.py]: @nashif any comments on this script?

File: doc/getting_started/installation_win.rst
Comment (carlescufi) [reply doc/getting_started/installation_win.rst]: So what's the conclusion here? I vote for having scripts/requirements-windows.txt` instead of having to type `windows-curses` but I don't mind if we keep `pip3 install --user -r scripts/requirements.txt windows-curses`

Comment (mbolivar) [reply doc/getting_started/installation_win.rst]: +1, though I think it should be `windows-requirements.txt` for parallelism with the existing `py2-requirements.txt` (i.e. keep the pattern that our requirements.txt files always end in `requirements.txt`, instead of being `requirements-xxx.txt` sometimes but not always).

**Comment by ulfalizer** [comment]
Show-all mode has been implemented:

ss9

For the search/jump-to window, I'm thinking of having an edit box at the top, with a list of matching symbol names appearing below it as you type. You could then step into the list with the up/down arrow keys, and jump to the selected symbol with `Enter`.

Show-all mode is needed since the symbol you want to jump to might not be visible. It will be enabled automatically in that case.

**Comment by ulfalizer** [comment]
Fixed some other minor stuff:

  - The save-and-quit dialog is now skipped if nothing has been changed

  - Bool/tristate symbol values can now be set by pressing `N`/`M`/`Y`, like in `make menuconfig`

  - The menu path at the top now scrolls to the right if needed to show the title of the current menu

Search should be next.

**Comment by ulfalizer** [comment]
Search and jump-to is coming along. Still some missing stuff to add and code to clean up.

Screenshot (interface description at bottom):

skarmbild fran 2018-05-04 07-58-21

I'm designing it so that you can (ideally) just type part of a symbol name and press Enter to jump to it.

Here's where you end up for the example above:

skarmbild fran 2018-05-04 07-58-45

The symbol wasn't currently visible, so show-all mode was turned on automatically. Red means currently invisible.

**Comment by ulfalizer** [comment]
Changes in new PR:

  - Added search and jump-to to the menuconfig implementation. It also supports regular expressions now:

    ss11

    ss12

  - Put the Windows requirements (with the extra `windows-curses` package) into a new `scripts/windows-requirements.txt` file.

    There's now a separate `scripts/common-requirements.txt` that lists packages needed on all systems. It's sourced by the system-specific requirements files.

    `scripts/requirements.txt` could be split into `scripts/linux-requirements.txt` and `scripts/mac-requirements.txt` as well, but I kept it for now. There are no Linux- or Mac-specific requirements yet.

  - Added back `diffconfig`

  - Added back note explaining how to list files starting with `.`

  - Fixed the wording issues in the documentation that were pointed out

**Review by carlescufi — APPROVED** [review]
I am happy with this, works fine on Windows. There are tweaks to the UI I would like to get in, but those are non-critical. All the comments I made have been addressed.

**Comment by carlescufi** [comment]
Tested on macOS as well with good results, +1

**Comment by SebastianBoe** [comment]
@ulfalizer : There are some remnants of mconf in the build system that should be removed.

```
cmake/host-tools.cmake
32:  KCONFIG_MCONF
```

File: cmake/kconfig.cmake
Comment (SebastianBoe) [inline cmake/kconfig.cmake]: We don't need this construct (RUNS_ON_WIN) now that everything runs on all platforms.

Comment (carlescufi) [reply cmake/kconfig.cmake]: We don't need it in `kconfig.cmake` but it still makes sense in other places, like for example QEMU?

**Comment by SebastianBoe** [comment]
This must be blocked until @dbkinder has had time to review the doc changes.

I didn't like some of the new phrasing, but was able to find better language.

File: cmake/kconfig.cmake
Comment (ulfalizer) [reply cmake/kconfig.cmake]: Yeah... `kconfig_target_list` is a bit silly now too.

Comment (SebastianBoe) [reply cmake/kconfig.cmake]: Correct, but this construct is not used in QEMU. So it is safe to remove it.

Comment (carlescufi) [reply cmake/kconfig.cmake]: Agreed, let's get rid of it

**Comment by carlescufi** [comment]
>This must be blocked until @dbkinder has had time to review the doc changes.

Fine with me, I thought he had already done a pass at it but I now see it was only me.

**Review by nashif — APPROVED** [review]

File: scripts/CMakeLists.txt
Comment (SebastianBoe) [inline scripts/CMakeLists.txt]: Now that conf is no longer a dependency we need to update the mac documentation that instructs users to build conf.

Comment (carlescufi) [reply scripts/CMakeLists.txt]: Nice catch. @ulfalizer can you get rid of the paragraph beginning with "Build Kconfig in :file:`$ZEPHYR_BASE/build` and add it to path" in  https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/getting_started/installation_mac.rst

Comment (carlescufi) [reply scripts/CMakeLists.txt]: Same for the paragraph beginning with "Within the MSYS console, build Kconfig in :file:`$ZEPHYR_BASE/build` and add it to path" in https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/getting_started/installation_win.rst

Comment (carlescufi) [reply scripts/CMakeLists.txt]: And same with the paragraph beginning with "On UNIX platforms, Build Kconfig in :file:`$ZEPHYR_BASE/build` and add it to path" in https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/getting_started/getting_started.rst. Also remove "Now that the host tools are installed,"

File: cmake/kconfig.cmake
Comment (ulfalizer) [reply cmake/kconfig.cmake]: Got rid of it.

File: scripts/CMakeLists.txt
Comment (ulfalizer) [reply scripts/CMakeLists.txt]: Fixed

**Review by SebastianBoe — APPROVED** [review]

File: doc/getting_started/installation_win.rst
Comment (carlescufi) [reply doc/getting_started/installation_win.rst]: I think there's no `if`, but rather `windows-curses; sys_platform == 'win32'`. See here:
https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers

I'm not against trying this out actually, it does keep things clean in terms of requirements files.

Comment (carlescufi) [reply doc/getting_started/installation_win.rst]: I am pretty sure this is safe actually. Python assigns `win32` to `sys.platform` on all Windows versions no matter the bitness of Python itself either, so I would go with this.

Comment (carlescufi) [reply doc/getting_started/installation_win.rst]: I tested this now on both Windows and Linux and can confirm it works fine. My vote is to go back to a single `requirements.txt` and use `windows-curses; sys_platform == 'win32'`

Comment (ulfalizer) [reply doc/getting_started/installation_win.rst]: Nice find.

I changed it. Looks like `sym_platform` might be reasonably safe.

File: doc/application/application.rst
Comment (dbkinder) [inline doc/application/application.rst]: The previous key explanations were "active", so let's make this one as well:

    #. Press :kbd:`?` to view information about the selected symbol.

File: doc/application/application.rst
Comment (dbkinder) [inline doc/application/application.rst]: Here too:

       Press :kbd:`ESC` or kbd:`Q` to return from the information display to the
       menu display.

File: doc/application/application.rst
Comment (dbkinder) [inline doc/application/application.rst]: The name now is simply ``macOS`` instead of ``Mac OS X``.

File: doc/application/application.rst
Comment (dbkinder) [inline doc/application/application.rst]: better is to write this as:

    .. visible (e.g., due to having unsatisfied dependencies) then ...

File: doc/application/application.rst
Comment (dbkinder) [inline doc/application/application.rst]: add some commas:

    jump to that symbol, in turn, to see

**Review by dbkinder — CHANGES_REQUESTED** [review]
Looks pretty good.  A few suggestions.

File: doc/application/application.rst
Comment (ulfalizer) [reply doc/application/application.rst]: Hmm... that makes it sound like something the user must do, since it's currently "step 4".

Think that's the real problem. I'll see if I can reorganize it.

File: doc/application/application.rst
Comment (ulfalizer) [reply doc/application/application.rst]: Got complaints when I added a comma after 'e.g.' in a different project. Maybe I should switch back. ;)

**Comment by ulfalizer** [comment]
@dbkinder 
Fixed.

Reorganized things slightly so that `?` isn't listed as a separate step. Check if it looks alright.

File: doc/application/application.rst
Comment (dbkinder) [reply doc/application/application.rst]: Ha. ;)  Most (but not all) writing style guides (e.g., Chicago Manual of Style, The Columbia Guide to Standard American English) recommend using a comma.  If you had used the phrase "for example" a comma would be expected.

**Review by dbkinder — APPROVED** [review]
Works for me, thanks!

**Pull request merged**
